### ISSUE TITLE
Mount overriding in java-microprofile stack corrupts extraction

### ISSUE THREAD

**Issue by gireeshpunathil** [appsody/stacks#280]
**Describe the bug**

https://github.com/appsody/stacks/blob/29d191bf76b8d59d22cb8a394ec1effb14673002/incubator/java-microprofile/image/Dockerfile-stack#L21

more precisely, token 3: `src:/project/user-app/src` is proving to be problematic. When extraction happens, the current directory is the parent of `src`, and because of this mount directive, the parent's mount is either ignored or abandoned - resulting in the project's configuration files being lost.

**To Reproduce**
1. appsody init java-microprofile
2. appsody extract -v
3. look at the extract location
4. ls -la on `user-app` 

**Expected behavior**
A complete copy of current location from where `appsody extract` was issued.

**Actual behaviour**

```
[root@linuxkit-025000000001 javam]# ls -la
total 16
drwxr-xr-x  7 root root  224 Aug 30 07:56 .
drwxr-xr-x 12 root root  384 Aug 30 08:04 ..
-rw-r--r--  1 root root   37 Aug 30 07:56 .appsody-config.yaml
-rw-r--r--  1 root root  271 Aug 30 07:56 .gitignore
-rwxr-xr-x  1 root root 7147 Aug 30 07:56 pom.xml
drwxr-xr-x  4 root root  128 Aug 30 07:56 src
drwxr-xr-x  4 root root  128 Aug 30 07:56 .vscode
[root@linuxkit-025000000001 javam]# ls -la /root/.appsody/extract/javam/user-app/
total 20
drwxr-xr-x 3 root root 4096 Aug 30 08:15 .
drwxr-xr-x 3 root root 4096 Aug 30 08:15 ..
-rwxr-xr-x 1 root root 7147 Aug 30 08:15 pom.xml
drwxr-xr-x 4 root root 4096 Aug 30 08:15 src
[root@linuxkit-025000000001 javam]# 
```

**Environment Details (please complete the following information):**
 - OS: Linux
 - Version : NA

**If applicable please specify:**
 - CLI version: `appsody vlatest`
 - Stack you are using: `java-microprofile`

**Additional context**
1. Was uncovered while testing for buildah.
2. This is the only mount token in all of the stacks that overlaps with the current location.

**Comment by gireeshpunathil** [comment]
In addition to the above, I see these inconsistencies:

1. Some stacks for example: `swift:0.1`, `nodejs-loopback:0.1`, `nodejs:0.2` have this mount:

`/:/project/user-app`

This is neither meaningful nor feasible while extracting: the ask is to mount the entire host's fs into the container? Is this intentional, and if yes, what purpose it solves? Are we hoping to capture many project files sprinkled in the host's system to be made available? or something else?

2. `java-spring-boot2:0.3` stack has this mount:

`.:/project/user-app`

this is fine, but the extractor takes care of this already. If this is setup at every stack, then the extractor logic could be simplified not to worry about that, and vice versa.

In short, for the extractor to function well, these incosistencies need to sortred out by defining a standard, as well as documenting it.

I am raising this as an issue (rather than a PR) because I am not 100% sure of other use cases of these MOUNT variables than for extraction, and guessing that I may be potentially missing something.

**Comment by neeraj-laad** [comment]
Thanks @gireeshpunathil for raising this issue. This is something that has always bothered me too.


To clarify, in both the cases we are mounting user's current workspace to the specified path in appsody container. So `/` is same as `.`. 

@appsody/cli team can one of you please explain what was the reason for using / instead of . for the mounts.

**Comment by neeraj-laad** [comment]
@skoh7645 Please can you test the stack by changing the mounts to use `.:/project/user-app`

**Comment by skoh7645** [comment]
@gireeshpunathil Sorry, why do you need every file to be copied? I would like to understand the use case better. We wrote the `extract` command as such to exclude certain files from being copied. This is to mimic what is inside the container, and create a clean project that is not polluted by files from the current workspace

**Comment by gireeshpunathil** [comment]
@skoh7645 - if you are alluding to the full mount (`/:/project/user-app`), then:
  - that mount is not being honored when I checked; can you please confirm (that it works as you wanted)?
 - assuming that it works, then we don't want any other mounts - because this covers the entire file system of the host!

**Comment by kylegc** [comment]
There's two topics of discussion in this issue:
### 1) Inconsistency between stacks using `/` and `.` in the mount var. 
I believe either one would will equally, so we should pick one and be consistent in all stacks.

@gireeshpunathil I think your confusion here is is that appsody prepends the current dir (the appsody project dir) path to the host mount before sending the mount to docker. For example, if pwd is `/home/user/appsodyproject` and the mount config is `/:/project/user-app`, the final mount sent to docker becomes `/home/user/appsodyproject/:/project/user-app`. Thus the entire filesystem won't be mounted, only the appsody project. This should probably be better documented.

### 2) Extract is not copying all the projects in the current directory. 
This is the topic original issue description. However, I believe it is working as designed. Extract is not meant to contain all the local configuration files, it should only contain the files required to build the final docker image. We gather these files by looking at what was mounted in the dev image.

@gireeshpunathil Please explain why you expect local configuration files to appear in extract.

**Comment by gireeshpunathil** [comment]
@kylegc - thanks for the explanations. Definitely, a document that explains the contracts will greatly help!

ok, so if the project configuration file is not required in the extraction, I am fine by that. But does that also mean nothing in the current project folder is required?

Let us look at what happens to 2 files `Foo.java` and `Foo.class` that I kept in the project folder - they are abandoned. Is this an expected behavior?

```
[root@j1 javam]# pwd
/root/javam
[root@j1 javam]# ls -la
total 28
drwxr-xr-x.  5 root root  147 Sep  5 01:56 .
dr-xr-x---. 12 root root 4096 Sep  5 01:54 ..
-rw-r--r--.  1 root root   37 Sep  5 01:54 .appsody-config.yaml
drwxr-xr-x.  4 root root  184 Sep  2 05:41 .appsody_init
-rw-r--r--.  1 root root    6 Sep  5 01:56 Foo.class
-rw-r--r--.  1 root root    6 Sep  5 01:56 Foo.java
-rw-r--r--.  1 root root  271 Sep  5 01:54 .gitignore
-rwxr-xr-x.  1 root root 7147 Sep  5 01:54 pom.xml
drwxr-xr-x.  4 root root   30 Sep  5 01:54 src
drwxr-xr-x.  2 root root   43 Sep  5 01:54 .vscode
[root@j1 javam]# appsody extract
Extracting project from development environment
Running command: docker[pull appsody/java-microprofile:0.2]
[Warning] The stack image does not contain APPSODY_PROJECT_DIR. Using /project
[Warning] Could not mount /root/.m2/repository:/mvn/repository because the local file was not found.
Running command: docker[create --name javam-extract -v /root/javam/src:/project/user-app/src -v /root/javam/pom.xml:/project/user-app/pom.xml appsody/java-microprofile:0.2]
Running command: docker[cp javam-extract:/project /root/.appsody/extract/javam]
Running command: docker[rm javam-extract -f]
Project extracted to /root/.appsody/extract/javam
[root@j1 javam]# ls -la /root/.appsody/extract/javam
total 44
drwxr-xr-x. 4 root root   184 Sep  2 05:41 .
drwxr-xr-x. 4 root root    30 Sep  5 01:56 ..
-rwxrwxr-x. 1 root root    33 Sep  2 05:39 .appsody-init.bat
-rwxrwxr-x. 1 root root    47 Sep  2 05:39 .appsody-init.sh
-rw-rw-r--. 1 root root   378 Sep  2 05:39 Dockerfile
-rw-rw-r--. 1 root root    42 Sep  2 05:39 .dockerignore
drwxr-xr-x. 3 root root    21 Sep  2 05:41 .mvn
-rwxrwxr-x. 1 root root 10078 Sep  2 05:41 mvnw
-rw-rw-r--. 1 root root  6609 Sep  2 05:41 mvnw.cmd
-rw-rw-r--. 1 root root  3618 Sep  2 05:39 pom.xml
drwxr-xr-x. 3 root root    32 Sep  5 01:56 user-app
-rwxrwxr-x. 1 root root  2963 Sep  2 05:39 validate.sh
[root@j1 javam]# ls -la /root/.appsody/extract/javam/user-app/
total 8
drwxr-xr-x. 3 root root   32 Sep  5 01:56 .
drwxr-xr-x. 4 root root  184 Sep  2 05:41 ..
-rwxr-xr-x. 1 root root 7147 Sep  5 01:54 pom.xml
drwxr-xr-x. 4 root root   30 Sep  5 01:54 src
[root@j1 javam]# 
```
Clearly, this is a side effect of deep mount (` /root/javam/src:/project/user-app/src`) superseding the outer mount (`/:/project/user-app`) and I find no way to reason this.

**Comment by kylegc** [comment]
Well, for this stack `/:/project/user-app` is not mounted:
https://github.com/appsody/stacks/blob/29d191bf76b8d59d22cb8a394ec1effb14673002/incubator/java-microprofile/image/Dockerfile-stack#L21

Thus, yes this is the expected behavior. The stack developers require you to put all code in the src dir for this stack.

**Comment by gireeshpunathil** [comment]
@kylegc - thanks for the explanations; I see your point now. Closing this, the only action is to improvement on the documentation ,specifically on the layout of the stack, what are  mounted and will be picked up for extraction and deployment thereon.

**Comment by kylegc** [comment]
FYI I created https://github.com/appsody/website/issues/233 to improve the documentation.

**Comment by neeraj-laad** [comment]
Thanks @kylegc 

Along with clearer documentation, we should favour `.` over `/` as that could avoid confusion in the first place.

**Comment by kylegc** [comment]
@appsody/stacks team what do you think of favoring `.` over `/` to start the mount path?

**Comment by kylegc** [comment]
@neeraj-laad should we update the existing stacks to use `.` then?

**Comment by ebullient** [comment]
For those that don't already use `.`, I think they should be updated. `.` is much more intuitive / understandable
